back/app/services/survey_results_generator_service.rb

```diff
@@ -107,70 +87,178 @@ def visit_file_upload(field)
     files = IdeaFile.where(id: file_ids).map do |file|
       { name: file.name, url: file.file.url }
     end
+    response_count = files.size
+    core_field_attributes(field, response_count).merge({
+      files: files
+    })
+  end
+
+  private
+
+  attr_reader :fields, :inputs, :locales
+
+  def core_field_attributes(field, response_count)
     {
       inputType: field.input_type,
       question: field.title_multiloc,
-      required: field.required,
-      totalResponses: files.size,
       customFieldId: field.id,
-      files: files
+      required: field.required,
+      grouped: false,
+      totalResponseCount: @inputs.count,
+      questionResponseCount: response_count
     }
   end
 
-  private
+  def visit_select_base(field, group_mode: nil, group_field_id: nil)
+    query = inputs
+    if group_field_id
+      if group_mode == 'user_field'
+        # Single user field grouped result
+        group_field = CustomField.find(group_field_id)
+        query = query.joins(:author)
+      else
+        # Single form field grouped result
+        group_field = find_question(group_field_id)
+      end
+      raise "Unsupported group field type: #{group_field.input_type}" unless group_field.input_type == 'select'
 
-  attr_reader :fields, :inputs, :locales
+      query = query.select(
+        select_field_query(field, as: 'answer'),
+        select_field_query(group_field, as: 'group')
+      )
+      answers = construct_grouped_answers(query, field, group_field)
+    else
+      query = query.select(
+        select_field_query(field, as: 'answer')
+      )
+      answers = construct_not_grouped_answers(query, field)
+    end
+
+    # Sort correctly
+    answers = answers.sort_by { |a| -a[:count] } unless field.input_type == 'linear_scale'
+    answers = answers.sort_by { |a| a[:answer] == 'other' ? 1 : 0 } # other should always be last
+
+    # Build response
+    build_select_response(answers, field, group_field)
```

Comment:
I'm a bit worried that this class has become too big and a bit hard to navigate and grasp. What would be a good chunk of logic that does some other work and so could be moved to a separate class? 

Maybe it could be a class that includes all the logic related to building select answers? And so includes construct_grouped_answers, construct_not_grouped_answers, build_select_response.

```suggestion
      SomeClassThatBuildsSelectResponse#construct_grouped_answers(query, field, group_field)
    else
      query = query.select(
        select_field_query(field, as: 'answer')
      )
      SomeClassThatBuildsSelectResponse#construct_not_grouped_answers(query, field)
    end
```

Reply:
Afraid I disagree on this unless we intend to reuse this elsewhere, which I don't think we do. I don't think it's unduly big at the moment but I'll keep this suggestion in mind for the separate performance improvements I want to do.